src/operator/controllers/intents_reconcilers/network_policy.go

```diff
@@ -463,13 +476,56 @@ func (r *NetworkPolicyReconciler) buildNetworkPolicyObjectForIntent(
 }
 
 func (r *NetworkPolicyReconciler) buildPodLabelSelectorFromIntent(intent otterizev1alpha2.Intent, intentsObjNamespace string) metav1.LabelSelector {
-	targetNamespace := intent.GetServerNamespace(intentsObjNamespace)
-	// The intent's target server made of name + namespace + hash
-	formattedTargetServer := otterizev1alpha2.GetFormattedOtterizeIdentity(intent.GetServerName(), targetNamespace)
+	otterizeServerLabel := map[string]string{}
+	if intent.IsTargetServerKubernetesService() {
+		formattedTargetServer := otterizev1alpha2.GetFormattedOtterizeIdentity(intent.GetTargetServerName(), intent.GetTargetServerNamespace(intentsObjNamespace))
+		otterizeServerLabel[fmt.Sprintf(otterizev1alpha2.OtterizeKubernetesServiceLabelKey, formattedTargetServer)] = "true"
+	} else {
+		targetNamespace := intent.GetTargetServerNamespace(intentsObjNamespace)
+		// The intent's target server made of name + namespace + hash
+		formattedTargetServer := otterizev1alpha2.GetFormattedOtterizeIdentity(intent.GetTargetServerName(), targetNamespace)
+		otterizeServerLabel[otterizev1alpha2.OtterizeServerLabelKey] = formattedTargetServer
+	}
+
+	return metav1.LabelSelector{MatchLabels: otterizeServerLabel}
+}
 
-	return metav1.LabelSelector{
-		MatchLabels: map[string]string{
-			otterizev1alpha2.OtterizeServerLabelKey: formattedTargetServer,
-		},
+func (r *NetworkPolicyReconciler) addPortRestrictionToNetworkPolicy(
+	ctx context.Context,
+	intent otterizev1alpha2.Intent,
+	intentsObjNamespace string,
+	policy *v1.NetworkPolicy) error {
+
+	svc := corev1.Service{}
+	kubernetesSvcName := intent.GetTargetServerName()
+	kubernetesSvcNamespace := intent.GetTargetServerNamespace(intentsObjNamespace)
+
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: kubernetesSvcNamespace,
+		Name:      kubernetesSvcName,
+	}, &svc)
+
+	if err != nil {
+		return err
+	}
+
+	portToProtocol := make(map[int]corev1.Protocol)
+	// Gather all target ports (target ports in the pod the service proxies to)
+	for _, port := range svc.Spec.Ports {
+		portToProtocol[port.TargetPort.IntValue()] = port.Protocol
```

Comment:
What if TargetPort is specified as string? We can note that it's unsupported here (you can provide the port's "name", and not the number, which would mean it's specified in the Pod)

Reply:
We can fetch the pod with the Service's label selector and convert the string to int. Tell me WDYT

Reply:
But then we would have to reconcile pods too, no? I think better to leave this as a limitation for now, and if there is no target port as int, send a Kubernetes Event Warning that no port was used because named ports are unsupported.